Alerting: Show legacy provisioned alert rules warning

gillesdemey opened the pull request:
**What is this feature?**

This PR adds a warning when a user is browsing a dashboard that has legacy alert rules embedded in the panel configuration.

A helpful message is included and a link to the documentation (link to documentation TBD) on how to proceed to migrate.

<img width="1202" alt="image" src="https://github.com/grafana/grafana/assets/868844/5aab5673-9e8c-482f-9c19-7c6606dc4fe2">

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/58169


**Notes for reviewers**

Setting this to draft mode for now – ideally this would only show for users of Grafana 11+ and I have to figure out a good way to do that (or just wait to merge until after 10.4)

Additionally the copy needs to be reviewed too before this is good to go (cc @brendamuir)

gillesdemey committed:
WIP

gillesdemey committed:
rest of the owl

brendamuir commented:
Hey Gilles -

konrad147 commented:
I wonder if we should display a similar warning on the `JSON model` page or `Provisioned dashboard` summary.
The Alert rules button is a new thing. Users might overlook it, and errors could go unnoticed. 

Most users modifying provisioned dashboards use the `Provisioned dashboard` summary view to copy the JSON content

![image](https://github.com/grafana/grafana/assets/6293563/b792eebf-9763-4528-8f40-5b61b22f7985)

![image](https://github.com/grafana/grafana/assets/6293563/01cf75ef-6776-4e01-a481-b0ef5e49c5e8)

gillesdemey commented:
Great idea! I'll see if I can move the fetching logic in to a wrapper component so we can have a portable component we can put anywhere we have a `dashboardUid`

gillesdemey committed:
refactor legacy alerts deprecation notice

gillesdemey committed:
add legacy alerts deprecation warning to dashboard JSON settings

gillesdemey committed:
omit pagination from interactive table if number of rows < page size

gillesdemey committed:
implement legacy alert notification for scenes

gillesdemey committed:
refactor

gillesdemey committed:
refactor to use dashboard type

gillesdemey commented on packages/grafana-ui/src/components/InteractiveTable/InteractiveTable.tsx:
```diff
@@ -179,7 +179,8 @@ export function InteractiveTable<TableData extends object>({
 
-  const paginationEnabled = pageSize > 0;
+  const multiplePages = data.length > pageSize;
+  const paginationEnabled = pageSize > 0 && multiplePages;
```
This changes seemed sensible for my use-case but maybe it was intended to also show the pagination controls even if the data set is smaller than the page size?

gillesdemey committed:
Merge remote-tracking branch 'origin/main' into alerting/show-legacy-provisioned-alert-rules-warning

gillesdemey committed:
trigger CI

gillesdemey committed:
update InteractiveTable test

gillesdemey committed:
Merge branch 'main' into alerting/show-legacy-provisioned-alert-rules-warning

gillesdemey committed:
only show warning to users using the new alerting

anyone wishing to stay on legacy alerting would probably not be interested to see the deprecation warning

gillesdemey committed:
simplify

ppcano commented:
This warning is shown whether the alert was previously migrated or not.

<img width="1397" alt="legacy_rules_in_dashboard" src="https://github.com/grafana/grafana/assets/825430/e54c5a9f-e6f5-4adb-99c2-b962578acb1a">

<br/><br/>

Clicking on the panel title will take the user directly to that panel, with the alert tab already open.


**1) Non-migrated alert**
  <img width="1704" alt="non_migrated_alert" src="https://github.com/grafana/grafana/assets/825430/5b870f98-b278-4a95-adc6-bf5032e6cdc6">


**2) Migrated alert**
<img width="1412" alt="migrated_alert" src="https://github.com/grafana/grafana/assets/825430/f1e3abc4-31ca-4262-b3cc-fb06ef8803dc">




The migrated alert case may be confusing:
- We inform the user that has legacy alerts but these were already migrated. 
- They may wonder, is the alert on the Alert tab the new or the old one?


I'm planning to document these various cases but wondered if there is anything else that could be done on the UI (cc @gillesdemey ).

soniaAguilarPeiron commented on public/app/features/alerting/unified/integration/AlertRulesDrawerContent.tsx:
```diff
@@ -27,12 +40,17 @@ export default function AlertRulesDrawerContent({ dashboardUid }: Props) {
     .flatMap((g) => g.rules)
     .filter((rule) => rule.annotations[Annotation.dashboardUID] === dashboardUid);
 
+  const loading = loadingRulesData;
```
why do we need this new variable `loadingRulesData`, why don't we use directly loading from here?:
`const { loading } = useAsync(async () => {`

konrad147 replied:
Should we include loading state of `fetchDashboard` too?

soniaAguilarPeiron reviewed (approved):
LGTM! 🚀
Added a non blocking comment 👀

konrad147 commented on public/app/features/alerting/unified/utils/misc.ts:
```diff
@@ -153,8 +153,22 @@ export function makeDashboardLink(dashboardUID: string): string {
+};
+export function makePanelLink(dashboardUID: string, panelId: string, queryParams: PanelLinkParams = {}): string {
+  const panelParams = new URLSearchParams({
+    viewPanel: panelId,
```
nit: Doesn't `...queryParams` do the job?

gillesdemey replied:
Oh that's a good point – I think I refactored this and forgot to simplify this part :D

konrad147 commented on public/app/features/alerting/unified/integration/LegacyAlertsDeprecationNotice.tsx:
```diff
@@ -0,0 +1,92 @@
+    return null;
+  }
+
+  return dashboard.uid ? <LegacyAlertsWarning dashboardUid={dashboard.uid} panels={panelsWithLegacyAlerts} /> : null;
```
Can `dashboard.uid` be undefined?

gillesdemey replied:
Yep, that's how it seems to be typed – I suspect the same model is used for existing dashboard as for dashboards that are being created and don't have a UID from the database yet.

gillesdemey committed:
PR feedback

konrad147 reviewed (approved):
LGTM!
